bin/lib/installable/edg.py

```diff
@@ -70,80 +150,41 @@ def _shim_default_shell(install_dir: Path, gcc: Path, **_kwargs) -> str:
-
-        unzip_dir = staging.path
-        unzip_dir.mkdir(exist_ok=True, parents=True)
-
-        def _call(checked_command):
-            _LOGGER.info("Running %s", shlex.join(checked_command))
-            subprocess.check_call(checked_command, cwd=unzip_dir)
-
-        if self._compiler_type == "default":
-            compiler_path = backend_path / "bin" / "gcc"
-        else:
-            compiler_path = backend_path / "bin" / self._compiler_type
-
-        def _query(lang: str, query_type: str) -> str:
-            command_to_run = [
-                self._scape_cmd,
-                f"--compiler-path={compiler_path}",
-                f"--lang={lang}",
-                self._compiler_type,
-                query_type,
-            ]
-            _LOGGER.info("Running %s", shlex.join(command_to_run))
-            return subprocess.check_output(command_to_run, cwd=unzip_dir).decode("utf-8").strip()
-
-        if self._compiler_type != "default":
-            with tempfile.NamedTemporaryFile() as temp_file:
-                amazon.s3_client.download_fileobj("compiler-explorer", f"opt-nonfree/{self._scraper}", temp_file)
-                temp_file.flush()
-                _call(["unzip", temp_file.name])
-                cpp_includes = _query("c++", "includes")
-                c_includes = _query("c", "includes")
-                version = _query("c", "version")
-
-            with tempfile.NamedTemporaryFile() as temp_file:
-                amazon.s3_client.download_fileobj("compiler-explorer", f"opt-nonfree/{self._macro_gen}", temp_file)
-                temp_file.flush()
-                command = ["bash", temp_file.name, f"--{self._compiler_type}", str(compiler_path)]
-                _LOGGER.info("Running %s", shlex.join(command))
-                subprocess.check_call(command, cwd=staging.path / self.untar_dir / self._macro_dir)
-        else:
-            cpp_includes = ""
-            c_includes = ""
-            version = ""
-
-        output_path = staging.path / self.untar_dir / "eccp-scripts"
-        output_path.mkdir(exist_ok=True)
-        script_path = output_path / f"eccp-{self._compiler_type}"
-        with script_path.open("w") as out:
-            out.write(
-                self._shim_shell_func(
-                    install_dir=self.install_context.destination / self.install_path,
-                    gcc=compiler_path,
-                    c_includes=c_includes,
-                    cpp_includes=cpp_includes,
-                    version=version,
-                )
-            )
-        script_path.chmod(0o755)
+
+        backend_compiler_path = _resolve_backend_compiler(self)
+        backend_compiler_scrape = _scrape_backend_compiler(self, staging, backend_compiler_path)
+
+        _write_predefined_macros(self, staging, backend_compiler_path)
+        _write_compiler_shim(self, staging, backend_compiler_path, backend_compiler_scrape)
```

Comment:
It _may_ be worth using two different types for `EdgCompilerInstallable` one that's `default` and one that's `non-default`.

`_write_predefined_macros` and `_scrape_backend_compiler` are effectively noops for `default` (and should only be necessary for GCC, Clang, and potentially Microsoft mode -- though I expect Microsoft mode/MSVC to be a bit more challenging/might need some unique handling of its own)